File: project/pyalg_api/model/antifraud/af_jac_base.py
# -*- coding: utf-8 -*-
from lib.application import db
from lib.logger import logger
import os
from .base import Base
from datetime import datetime, timedelta
from sqlalchemy import and_

class AfJacBase(db.Model, Base):
    # 指定数据库
    __bind_key__ = 'xhh_antifraud'
    __tablename__ = 'af_jac_base'

    id = db.Column(db.BigInteger, primary_key=True)
    aid = db.Column(db.Integer,nullable=False)
    jac_match_id = db.Column(db.BigInteger, nullable=False, index=True)
    request_id = db.Column(db.BigInteger, nullable=False, index=True)
    base_id = db.Column(db.BigInteger, nullable=False, index=True)
    user_id = db.Column(db.BigInteger, nullable=False, index=True)
    loan_id = db.Column(db.BigInteger, nullable=False, index=True)
    mobile = db.Column(db.String(32), nullable=False, index=True)
    jac_status = db.Column(db.Integer, nullable=False)
    base_status = db.Column(db.Integer, nullable=False)
    create_time = db.Column(db.DateTime, nullable=False, index=True)
    modify_time = db.Column(db.DateTime, nullable=False)

    def getById(self, id):
        db_base = self.query.get(int(id))
        return db_base

    def save(self):
        db.session.add(self)
        try:
            db.session.flush()
            db.session.commit()
        except Exception:
            db.session.rollback()
            raise

    def getJaccardData(self, start_time):
        '''
        获取需要匹配间接关系(jaccard)的数据
        '''
        where = and_(
            AfJacBase.create_time >= start_time,
            AfJacBase.jac_status == 0
        )

        res = db.session.query(AfJacBase) \
            .filter(where) \
            .order_by(AfJacBase.create_time) \
            .limit(500) \
            .all()

        return res
    def getRelationData(self, start_time):
        '''
        获取已存在关系(relation)的数据
        '''
        where = and_(
            AfJacBase.create_time >= start_time,
            AfJacBase.jac_status == 2
        )

        res = db.session.query(AfJacBase) \
            .filter(where) \
            .order_by(AfJacBase.create_time) \
            .limit(500) \
            .all()
        return res
    def getJaccardDataByids(self, ids):
        '''
        获取需要匹配间接关系(jaccard)的数据
        '''
        where = and_(
            AfJacBase.jac_status == 0
        )
        res = db.session.query(AfJacBase) \
            .filter(AfJacBase.match_status.in_([0,1,2])) \
            .order_by(AfJacBase.create_time) \
            .group_by(AfJacBase.user_id) \
            .filter(AfJacBase.user_id.in_(ids)) \
            .limit(500) \
            .all()
        return res

    def getAfBaseData(self, start_time):
        where = and_(
            AfJacBase.create_time >= start_time,
            AfJacBase.jac_status == 0, #初始
            # AfBase.aid == aid
        )

        res = db.session.query(AfJacBase)   \
            .filter(where) \
            .order_by(AfJacBase.create_time) \
            .limit(1000) \
            .all()

        return res

    def lockMatchStatus(self, data, status):
        if len(data) == 0:
            return False

        now = datetime.now()
        for o in data:
            o.modify_time = now
            o.jac_status = status  # 锁定 | 完成

        db.session.commit()

        return True

    def finishBase(self, base_id):
        '''
        反欺诈分析结束更新base_id = 2
        '''
        res = db.session.query(AfJacBase)\
            .filter(AfJacBase.base_id == base_id)\
            .update({AfJacBase.base_status:2,AfJacBase.modify_time:datetime.now()})
        if res == 0:
            logger.error("af_base.id:%s update af_jac_base is fail" % (base_id))
            return False
        db.session.commit()
        return True

    def finishJac(self, match_status):
        '''
        杰卡德关系结束匹配流程
        '''
        self.match_status = match_status
        self.modify_time = datetime.now()
        db.session.commit()

File: project/cloud/models/anti/AfJacBase.php
<?php

namespace app\models\anti; 

use Yii; 

/** 
 * This is the model class for table "af_jac_base". 
 * 
 * @property string $id
 * @property string $request_id
 * @property integer $aid
 * @property string $user_id
 * @property string $loan_id
 * @property string $mobile
 * @property integer $base_status
 * @property integer $jac_status
 * @property string $create_time
 * @property string $modify_time
 */ 
class AfJacBase extends AntiBaseModel
{ 
    const INIT = 0;
    const DOING = 1;
    const FINISHED = 2;
    /** 
     * @inheritdoc 
     */ 
    public static function tableName() 
    { 
        return 'af_jac_base'; 
    }

    /** 
     * @inheritdoc 
     */ 
    public function rules() 
    { 
        return [
            [['jac_match_id', 'request_id', 'aid', 'user_id', 'base_id','loan_id', 'base_status', 'jac_status'], 'integer'],
            [['mobile', 'base_status', 'jac_status', 'create_time', 'modify_time'], 'required'],
            [['create_time', 'modify_time'], 'safe'],
            [['mobile'], 'string', 'max' => 32]
        ]; 
    } 

    /** 
     * @inheritdoc 
     */ 
    public function attributeLabels() 
    { 
        return [ 
            'id' => 'ID',
            'jac_match_id' => 'af_jac_matcah表主键',
            `request_id` => '请求处理id',
            'aid' => '业务ID',
            'base_id' => 'af_base表主键ID',
            'user_id' => '用户ID',
            'loan_id' => '贷款ID',
            'mobile' => '用户手机号',
            'base_status' => '反欺诈数据分析状态:0:初始; 1:锁定; 2:成功',
            'jac_status' => '杰卡德关系状态:0:初始; 1:锁定; 2:成功',
            'create_time' => '创建时间',
            'modify_time' => '修改时间',
        ]; 
    } 

    /**
     * 获取需要匹配间接关系的数据
     */
    public function getJaccardData($data, $field = '*')
    {
        if (empty($data)) {
            return false;
        }
        $where = [
            'AND',
            ['jac_status' => $data['jac_status']],
            ['>', 'create_time', $data['create_time']],
        ];
//        $query = static::find()->where($where)->select($field)->asArray()->limit(500)->orderBy('create_time desc');
//        echo $query->createCommand()->getRawSql(); exit;
        return static::find()->where($where)->select($field)->asArray()->limit(500)->orderBy('create_time desc')->all();
    }

    /**
     * 锁定为间接关系处理中状态
     */
    public function lockJcards($ids) {
        if (empty($ids)) {
            return false;
        }
        $sets = [
            'jac_status' => self::DOING,
            'modify_time' => date('Y-m-d H:i:s'),
        ];
        $where = [
            'id' => $ids,
            'jac_status' => self::INIT,
        ];
        $result = static::updateAll($sets, $where);
        return $result;
    }

    /**
     * 更新为结束状态
     */
    public function finishJcard($id,$jac_match_id) {
        if (empty($id)) {
            return false;
        }
        $sets = [
            'jac_status' => self::FINISHED,
            'modify_time' => date('Y-m-d H:i:s'),
            'jac_match_id' => $jac_match_id,
        ];
        $where = [
            'id' => $id
        ];
        $result = static::updateAll($sets, $where);
        return $result;
    }
}
